### PR TITLE
364 Content Negotiation

### LINKED ISSUE #364: Javascript / Angular Rich Client SPA Support
Ensure Servlet/Spring/Spring Boot/Spring Security endpoints are rich-client-compatible per framework spec [here](https://github.com/stormpath/stormpath-framework-spec/blob/master/single-page-apps.md):

* html5 push state mode that will always render a static html file on GETs
* POSTs need to support both form-url-encoded content (server-rendered forms) and JSON (ria/spa)
* 3rd party login providers should display if in the Application's account store mappings

Note: This issue depends on implementing the `application/json` model responses for `/login` and `/register` in [this](https://github.com/stormpath/stormpath-sdk-java/issues/577) issue and [this](https://github.com/stormpath/stormpath-sdk-java/issues/583) issue, respectively.

### PR DESCRIPTION
**Pull request by mrioan**
#How to UAT
[This workflow](https://github.com/stormpath/stormpath-framework-spec/blob/master/requests.md#content-type-negotiation) should be working as described. Servlet, Spring, Spring Boot and Spring Security should be tested.

Please also check how registration and login (POST) is working with JSON.

Resolves #364 (some parts of it)

### CODE REVIEW

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/mvc/ControllerFilter.java [hunk 1]
```diff
@@ -116,6 +120,22 @@ protected void redirect(HttpServletRequest request, HttpServletResponse response
...
         View view = this.viewResolver.getView(vm, request);
-        view.render(request, response, vm);
+        if (view != null) {
+            view.render(request, response, vm);
+        }
+        response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+    }
+
+    /** @since 1.0.0 */
+    protected List<MediaType> prroducesMediaTypes() {
```
Comment (mraible) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/mvc/ControllerFilter.java]: Typo in method name: `prroducesMediaTypes` should be `producesMediaTypes`.

File: extensions/spring/boot/stormpath-thymeleaf-spring-boot-starter/src/main/resources/templates/stormpath/login.html [hunk 2]
```diff
@@ -16,6 +16,7 @@
 <html xmlns:th="http://www.thymeleaf.org">
 <head>
     <title th:text="#{stormpath.web.login.title}">|Login|</title>
+    <!--/*/ <th:block th:include="${headViewName} :: ${headFragmentSelector}"/> /*/-->
```
Comment (mraible) [inline extensions/spring/boot/stormpath-thymeleaf-spring-boot-starter/src/main/resources/templates/stormpath/login.html]: I'd remove this commented out line. You can always use Git's history to find it again if you need it.

File: extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java [hunk 3]
```diff
@@ -323,12 +324,12 @@ public Controller stormpathLoginController() {
...
-    @Override
-    public Controller stormpathSpaController() {
-        return super.stormpathSpaController();
-    }
+//    @Bean
+//    @ConditionalOnMissingBean(name="stormpathSpaController")
+//    @Override
+//    public Controller stormpathSpaController() {
+//        return super.stormpathSpaController();
+//    }
```
Comment (mraible) [inline extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/main/java/com/stormpath/spring/boot/autoconfigure/StormpathWebMvcAutoConfiguration.java]: Remove commented code.

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java [hunk 4]
```diff
@@ -1,75 +1,91 @@
...
-    private final Controller spaViewController;
-
     private final String jsonView;
 
-    private final List<MediaType> producedMediaTypes;
+    private final List<MediaType> producesMediaTypes;
 
-    public SpringSpaController(Controller delegate, Controller spaViewController, String jsonView, List<MediaType> producedMediaTypes) {
+    public SpringSpaController(Controller delegate,
+                               //Controller spaViewController,
```
Comment (mraible) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java]: Remove commented code.

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java [hunk 5]
```diff
@@ -1,75 +1,91 @@
...
+                               List<MediaType> producesMediaTypes) {
         Assert.notNull(delegate, "Delegate controller cannot be null.");
-        Assert.notNull(spaViewController, "SPA view controller cannot be null.");
-        Assert.notEmpty(producedMediaTypes, "produced media types cannot be null or empty.");
+        //Assert.notNull(spaViewController, "SPA view controller cannot be null.");
+        Assert.notEmpty(producesMediaTypes, "produced media types cannot be null or empty.");
         Assert.hasText(jsonView, "jsonView cannot be null or empty.");
         this.delegate = delegate;
-        this.spaViewController = spaViewController;
+        //this.spaViewController = spaViewController;
```
Comment (mraible) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java]: Remove commented code.

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java [hunk 6]
```diff
@@ -1,75 +1,91 @@
...
         this.jsonView = jsonView;
-        this.producedMediaTypes = producedMediaTypes;
+        this.producesMediaTypes = producesMediaTypes;
     }
 
     @Override
     protected ModelAndView handleRequestInternal(HttpServletRequest request, HttpServletResponse response) throws Exception {
 
-        Controller controller = delegate;
+        //Controller controller = delegate;
```
Comment (mraible) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java]: Remove commented code.

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java [hunk 7]
```diff
@@ -1,75 +1,91 @@
...
 
-        String method = request.getMethod();
+        //String method = request.getMethod();
```
Comment (mraible) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/mvc/SpringSpaController.java]: Remove commented code.

**Comment by mraible** [comment]
For UAT, it seems like it would be best to write an integration test that tests all the endpoints that have JSON/HTML support. Are the endpoints and expected parameters documented somewhere?

**Comment by lhazlewood** [comment]
We have a TCK that should test these endpoints for both HTML and JSON. @mrioan have you been able to run that yet?

**Comment by dogeared** [comment]
@lhazlewood - there are a lot of overlapping dependencies. Before we can reliably use the TCK, we need to get `/register` (#631) and `/login` (#623) merged. We are going to be discussing this coordination of merges today.

**Pull request merged**